docs: migration guide (#70)
Unleash/unleash-android: docs/MigrationGuide.md

# Migration Guide from unleash-android-proxy-sdk

This guide provides detailed steps for migrating your project from the Unleash Android Proxy SDK version to the newer Unleash Android SDK.

> We will focus on the [previous sample application](https://github.com/Unleash/unleash-android-proxy-sdk/tree/main/samples/android) and specifically highlighting the changes from this pull request: https://github.com/Unleash/unleash-android-proxy-sdk/pull/83

Reviewer: Not really sure what you mean here by "highlighting the changes". Is it that:
- we've already got a sample application for the old proxy sdk
- the pull request makes some changes to make it compatible with the new sdk?

```suggestion
We will focus on the [previous sample application](https://github.com/Unleash/unleash-android-proxy-sdk/tree/main/samples/android), specifically highlighting the changes from this pull request: https://github.com/Unleash/unleash-android-proxy-sdk/pull/83
```

## Benefits of Migrating

This version of the Unleash Android SDK introduces several improvements, including:

> - No need for SLF4J dependency and the usage of the native Android logging system.

Reviewer: It's unclear whether it's "no need for [...] the usage of the native Android logging system", or not. I suspect that the android logging system replaces SLF4J, though. How about this?

```suggestion
- No need for the SLF4J dependency. Instead, use the native Android logging system.
```

or

```suggestion
- Uses the native Android logging system instead of SLF4J.
```

> - Respecting the Android lifecycle to stop polling and sending metrics in the background.

Reviewer: ```suggestion
- Respects the Android lifecycle and stops polling and sending metrics in the background.
```

> - The new SDK respects the minimum Android API level 21, but we recommend API level 23.

Reviewer: This is just for consistency

```suggestion
- Respects the minimum Android API level 21, but we recommend API level 23.
```

> - Monitoring network connectivity to avoid unnecessary polling (requires API level 23 or above).

Reviewer: ```suggestion
- Monitors network connectivity to avoid unnecessary polling (requires API level 23 or above).
```

## Overview

The new SDK introduces several changes and improvements, including API modifications and a new initialization process.

## Step-by-Step Migration

### 1. Update Gradle Dependency

First, update the dependency in your `build.gradle` file:

```gradle
dependencies {
    // Remove the old SDK
    implementation 'io.getunleash:unleash-android-proxy-sdk:0.5.0'

    // Add the new SDK
    implementation 'io.getunleash:unleash-android:$version'
}
```

### 2. Update the initialization code

> We won't cover all the details here as most of the configuration can be set using the builders fluent methods. However, the main difference is that the new SDK requires an `Application` context to be passed to the `Unleash` constructor. This is necessary to monitor the network connectivity and respect the Android lifecycle (usually it can be injected if using hilt with `@ApplicationContext context`).

Reviewer: That last parenthesis is a little jumbled, but I lack a lot of context. What is hilt? Some DI system? How about unpacking it into its own sentence and giving it some more room to breathe? Something like this
```suggestion
We won't cover all the details here as most of the configuration can be set using the builders fluent methods. However, the main difference is that the new SDK requires an `Application` context to be passed to the `Unleash` constructor. This is necessary to monitor the network connectivity and respect the Android lifecycle. If you use hilt, this can be injected with `@ApplicationContext context`.
```

 
#### Unleash context initialization
The main differences are:

> 1. The application name is no longer configurable through the context as it is unique across the application. The `appName` should be set using the `UnleashConfig` builder.

Reviewer: I'm not sure unique is the right word here. What about ... static? constant?

```suggestion
1. The application name is no longer configurable through the context, as it is constant throughout the application's lifetime. The `appName` should be set using the `UnleashConfig` builder.
```

3. Update the import statements to use the new SDK classes.

```kotlin
val unleashContext = UnleashContext.newBuilder()
    // .appName("unleash-android") // remove this line
    // .instanceId("main-activity-unleash-demo-${Random.nextLong()}") // remove this line
    .userId("unleash_demo_user")
    .sessionId(Random.nextLong().toString())
    .build()
```

#### Unleash configuration
The main differences are:
1. Metrics are enabled by default.
2. App name is now a mandatory parameter to the builder.
3. Instance id is no longer configurable.
4. The polling mode is now a polling strategy with a fluent API.
5. The metrics interval is now part of the metrics strategy with a fluent API.

**Old SDK**
```kotlin
UnleashConfig.newBuilder()
    .appName("unleash-android")
    .instanceId("unleash-android-${Random.nextLong()}")
    .enableMetrics()
    .clientKey("xyz")
    .proxyUrl("https://eu.app.unleash-hosted.com/demo/api/frontend")
    .pollingMode(
        PollingModes.autoPoll(
            autoPollIntervalSeconds = 15
        ) {

        }
    )
    .metricsInterval(5000)
    .build()
```

**New SDK**
```kotlin
UnleashConfig.newBuilder("unleash-android")
    .clientKey("xyz")

>     .proxyUrl("https://eu.app.unleash-hosted.com/demo/api/frontend")

Reviewer: do we still use `proxyUrl`? What if you use edge? Or Unleash directly (as in this example)? Is it too late to rename that property?

Author: I think for consistency with other sdks proxyUrl is the right thing here. Maybe in the future we could replace this, but I'm not sure if edge is best, I think I rather use frontend or frontendApiUrl... but I want to keep that out of this PR otherwise we keep having SDKs that drift apart.

Author: I wasn't sure about this. I suspect that 5 levels of headers is quite a lot for such short guide. Maybe I should organize this differently but the bold text seems to be fine IMO

Reviewer: Yes , sir 🫡

    .pollingStrategy.interval(15000)
    .metricsStrategy.interval(5000)
    .build()
```

#### Creating the Unleash instance
The previous SDK used a builder to construct the Unleash instance while the new SDK relies on constructor parameters. There are also other meaningful changes:

1. The new SDK does not start automatically. You need to call `unleash.start()` to start the polling and metrics collection.
2. The new SDK accepts event listeners at the constructor level or as parameters when calling `unleash.start()` (you can also edit your config object setting `delayedInitialization` to false).
3. The interface `UnleashClientSpec` is now `Unleash`.

> 

Reviewer: Think you forgot a header here

```suggestion

**Old SDK**
```

Author: I wasn't sure about this. I suspect that 5 levels of headers is quite a lot for such short guide. Maybe I should organize this differently but the bold text seems to be fine IMO

Reviewer: Oh, sorry, it's not an actual header, though, just bold text. But my point is that you've marked all the other code samples with either `New SDK` or `Old SDK`. I think you just forgot this one.

```kotlin
UnleashClient.newBuilder()
    .unleashConfig(unleashConfig)

>     .cache(InMemoryToggleCache()) // This was optional in the previous SDK and in the new one so we are ignoring it

Reviewer: I don't understand what the comment means. Why are you ignoring it?

Author: I was just trying to stay close to the code of the sample app that was including this cache implementation. But there's no point in doing that as the default implementation is the same (if you don't set cache). So I thought just commenting that in the code but I also didn't quite like it. I'll just remove this line
